deploy-agent/deployd/common/helper.py

```diff
@@ -65,17 +65,26 @@ def get_stale_builds(build_timestamps, num_builds_to_retain=2):
     def clean_package(base_dir, builds, build_name):
         local_fn = '{}-{}.*'.format(build_name, builds)
         try:
+            # Remove extracted pointer from disk
             extracted_file = os.path.join(base_dir, '{}.extracted'.format(builds))
             if os.path.exists(extracted_file):
                 os.remove(extracted_file)
+            # Remove staged pointer from disk
             staged_file = os.path.join(base_dir, '{}.staged'.format(builds))
             if os.path.exists(staged_file):
                 os.remove(staged_file)
         except OSError as e:
             log.error(e)
 
         try:
+            # Remove build directory from disk
             shutil.rmtree(os.path.join(base_dir, builds))
+        except BaseException as e::
+            # Catch base exception class, as there's a mulitude of reasons a rmtree can fail
+            log.error(e)
+
+        try:
```

Comment:
Since we only log exceptions in this function, how would you think of merging all the try's into one try, and simplify the exception catching process?

Reply:
Correct me if I'm wrong, but if we put all of the logic into a single exception handler, if the first function fails, the rest will not be executed - thus if trying to remove the `<id>.staging` file fails, none of the other cleanup functions will execute. 

Yes, having a bunch of exception handlers isn't pretty, but it ensures that an early failure won't leave even more artifacts laying around.

Reply:
I would also be in favor of catching earlier OSErrors and then any rmtree errors separately as to actually execute the remove if that if statement is reached.

If casting such a wide net, being redundant/explicit is not necessarily a bad thing.

Reply:
You're right that we want to catch all the errors. One option to make it clean is to define a list of functions, and then iterate on those function with one try-and-except, e.g.
for func in func_list:
  try: 
     func()
  except 
    log.exception

Reply:
Yes, you're right that we want to catch all the exceptions and give all the functions a chance to run.

One way to make it work is using a list of functions, e.g.
func_list = [func1, func2, func3]
for func in func_list:
    try: 
        func
    except IOError:
        xxx
    except Exception:
        xxx

Not sure whether it's worth it :-)